src/librustc_mir/transform/const_prop.rs

```diff
@@ -518,18 +518,19 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         }
     }
 
-    fn check_unary_op(&mut self, arg: &Operand<'tcx>, source_info: SourceInfo) -> Option<()> {
+    fn check_unary_op(
```

Comment:
It seems a bit weird to me that we added an assert here for `op == UnOp::Neg` but at the same time removed the checking at this function's callsite for `Neg`. Perhaps we should check in the body of this function that `op == UnOp::Neg` and if it isn't early return `Some(())` since as the assert mentions, "Neg is the only UnOp that can overflow". This function doesn't do anything else so that's the only `UnOp` we care about.

Reply:
I basically moved the "special hack" as far down as I could. Now the caller (that used to check for `Neg`) no longer needs to know which operations can overflow, it just always asks the Miri engine for this.

But when reporting an error, all we got is `OverflowNeg`, so we have to check that it really was a `Neg`.

> Perhaps we should check in the body of this function that op == UnOp::Neg and if it isn't early return Some(()) since as the assert mentions, "Neg is the only UnOp that can overflow". This function doesn't do anything else so that's the only UnOp we care about.

The point of this was not to encode any Miri-engine-internal assumptions here. If, say, some other unary op started to overflow, I want this to still work as good as it can. With your proposal it would silently discard that overflow. With my proposal it would ICE, which is better as we'll notice that.

The consistent think to do would be to have something like what we have for `BinOp`: a panic/assertion constructor that works for any unary operator. But that might be weird elsewhere.